Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs

```diff
@@ -54,34 +53,103 @@ public override bool CanConvert(Type objectType)
-
-            var childTypes = GetChildTypes(objectType);
+            if (reader.TokenType == JsonToken.Null)
+            {
+                return null;
+            }
 
-            var testObjects = childTypes.Select(Activator.CreateInstance)
-                .Select(instance => new { Instance = instance, SerializedInstance = Serialize(serializer, instance) })
-                .ToList();
+            var jsonObject = JObject.Load(reader);
+            var childTypes = GetChildTypes(objectType).ToList();
 
-            var testKeys = testObjects.ToDictionary(
-                to => to,
-                to => GetKeys(to.SerializedInstance).Select(kvp => kvp.Key).ToList());
+            var jsonProperties = GetProperties(jsonObject);
+            var candidateChildTypes = new List<Type>();
+            foreach (var childType in childTypes)
+            {
+                var childTypeProperties = childType.GetProperties().Select(t => t.Name).ToList();
+                var childTypeFields = childType.GetFields().Select(t => t.Name).ToList();
+                var childTypePropertiesAndFields = new HashSet<string>(
+                    childTypeProperties.Concat(childTypeFields),
+                    StringComparer.OrdinalIgnoreCase);
+                if (jsonProperties.All(p => childTypePropertiesAndFields.Contains(p)))
+                {
+                    candidateChildTypes.Add(childType);
+                }
+            }
 
-            var target = testObjects.Select(to => new { TestObject = to, Keys = testKeys[to] })
-                .OrderBy(t => t.Keys.Count)
-                .FirstOrDefault(t => !sourceKeys.Except(t.Keys).Any()); // Any keys in source that are not in the target object?
+            var deserizliedObjects = new List<object>();
+            foreach (var candidateChildType in candidateChildTypes)
+            {
+                try
+                {
+                    var deserializedObject = serializer.Deserialize(jsonObject.CreateReader(), candidateChildType);
+                    if (deserializedObject != null)
+                    {
+                        deserizliedObjects.Add(deserializedObject);
+                    }
+                }
+                catch (JsonException)
+                {
+                }
+            }
 
-            if (target == null)
+            if (deserizliedObjects.Count == 0)
             {
                 throw new JsonSerializationException(
                     string.Format(CultureInfo.InvariantCulture, "Unable to deserialize to type {0}, value: {1}", objectType, jsonObject));
             }
 
-            serializer.Populate(jsonObject.CreateReader(), target.TestObject.Instance);
+            if (deserizliedObjects.Count > 1)
+            {
+                var matchingTypes =
+                    deserizliedObjects.Select(_ => _.GetType().Name)
+                        .Aggregate((running, current) => running + " | " + current);
```

Comment:
Won't this do " | FirstType | SecondType | ... | LastType"? I'd prefer to not have additional characters at the beginning or the end.